vms/evm/uptimetracker/validator.go

```diff
@@ -0,0 +1,236 @@
+// Copyright (C) 2019-2025, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package uptimetracker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ava-labs/libevm/log"
+
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow"
+	"github.com/ava-labs/avalanchego/snow/uptime"
+	"github.com/ava-labs/avalanchego/snow/validators"
+	"github.com/ava-labs/avalanchego/utils/set"
+	"github.com/ava-labs/avalanchego/utils/timer/mockable"
+)
+
+const (
+	// SyncFrequency is the recommended frequency for calling Sync()
+	// VMs can use this as a default but are free to choose their own timing
+	SyncFrequency = 1 * time.Minute
+)
+
+// The UptimeTracker struct is responsible for managing the state of validators by fetching
+// information from P-Chain state (via GetCurrentValidatorSet in chain context) and updating
+// the local state accordingly. The caller is responsible for calling Sync() periodically to
+// keep the state up-to-date. The sync operation first removes validators that are no longer
+// in the P-Chain validator set, then adds new validators and updates existing validators.
+// This order of operations ensures that uptimes of validators being removed and re-added
+// under the same nodeIDs are updated in the same sync operation despite having different
+// validationIDs.
```

Comment:
This comment is so verbose + leaky. Why do we care about things like:
> The sync operation first removes validators that are no longer
// in the P-Chain validator set, then adds new validators and updates existing validators.
// This order of operations ensures that uptimes of validators being removed and re-added
// under the same nodeIDs are updated in the same sync operation despite having different
// validationIDs.

This comment is just reciting the implementation of `Sync` (so it doesn't even belong here, it belongs on `Sync` - and even there it would still be leaky).

I think a struct-level comment is appropriate here, but we should aim to have something that tells us what this type's responsibility is - not what the code does at a line-by-line-level.